packages/focal/src/react/react.ts

```diff
@@ -15,111 +15,113 @@ export interface Subscription {
+    [K in keyof T]: T[K] | Observable<T[K]>
+  }
 
-export interface LiftWrapperProps<TProps> {
-  component: React.Component<TProps, any>
-    | React.StatelessComponent<TProps>
+  export interface Props<TProps> {
+    component: React.Component<TProps, any>
+    | React.FunctionComponent<TProps>
     | React.ComponentClass<TProps>
     | React.ComponentType
     | keyof React.ReactHTML
-  props: Lifted<TProps>
-}
+    props: Lifted<TProps>
+  }
 
-export interface LiftWrapperState {
-  renderCache?: React.DOMElement<any, any> | null
-  subscription?: Subscription | null
-}
+  export type RenderCache = React.DOMElement<any, any> | null
+  export type WrapperSubscription = Subscription | null
 
-/**
- * A wrapper component that allows to use observables for prop values of a
- * given component.
- */
-export class LiftWrapper<TProps>
-    extends React.Component<LiftWrapperProps<TProps>, LiftWrapperState> {
-  state = LiftWrapper._initState
+  export class Renderer<TProps> extends React.PureComponent<Props<TProps>> {
+    static displayName = 'LiftWrapper'
 
-  static _initState: LiftWrapperState = {
-    renderCache: null,
-    subscription: null
-  }
+    subscription: WrapperSubscription = null
+    renderCache: RenderCache = null
 
-  static _endState: LiftWrapperState = {
-    subscription: null
-  }
+    // variable to track sync emit from observable in Render{One, Many}.
+    // prevent waste render during initial render & re-subscribe
+    isSubscribed = false
 
-  render() {
-    return this.state.renderCache || null
-  }
+    // eslint-disable-next-line camelcase
+    UNSAFE_componentWillMount() {
+      this._subscribe(this.props)
+    }
 
-  private _subscribe(newProps: LiftWrapperProps<TProps>) {
-    const { props, component } = newProps
+    // eslint-disable-next-line camelcase
+    UNSAFE_componentWillReceiveProps(nextProps: Props<TProps>) {
+      this._unsubscribe()
+      this._subscribe(nextProps)
+    }
 
-    let n = 0
-    // eslint-disable-next-line @typescript-eslint/no-use-before-define
-    walkObservables(props, () => n += 1)
+    componentWillUnmount() {
+      this._unsubscribe()
+      this.setSubscription(null)
+    }
 
-    switch (n) {
-      case 0:
-        this.setState({
-          subscription: null,
-          // eslint-disable-next-line @typescript-eslint/no-use-before-define
-          renderCache: render(component, props)
-        })
-        break
-
-      // @NOTE original Calmm code below
-      // The created object is never used and it looks like that
-      // the useful work is done in the constructor.
-      // Could this be replaced by a regular closure? Perhaps using
-      // a class is an optimization?
-      case 1:
-        new RenderOne(this, newProps) // eslint-disable-line
-        break
-      default:
-        new RenderMany(this, newProps, n) // eslint-disable-line
-        break
+    setRenderCache(cache: RenderCache): void {
+      if (this.renderCache === cache) {
```

Comment:
Does it make sense to check equality with `!structEq(this.renderCache, cache)` as well? Just prevent rendering of the same state twice? This probably should be handled by React itself, but for some reason, we do that in the latest version. https://github.com/grammarly/focal/pull/82/files?diff=split&w=1#diff-71010d1ee0e73a76461ec411a47605c1207b1982317b93bd4fdc7c35fb7c6b1fL403